Can support be provided for truenas scale?

TrueNAS is an open source network storage system on which I run my workloads, using the latest TrueNAS SCALE 23.10.2 version.
It now has a 6.6 kernel, but I don't know what to do with it.
https://github.com/truenas/linux.git
Provide the command install-dev-tools to enable developer mode and make it writable.
Download the install.sh script and run it locally, delete linux-header-(uname -a). Then run it, but the following error occurs



```Error! Bad return status for module build on kernel: 6.6.44-production+truenas (x86_64)
Consult /var/lib/dkms/dxgkrnl/149cbd13f/build/make.log for more information.
```

```
DKMS make.log for dxgkrnl-149cbd13f for kernel 6.6.44-production+truenas (x86_64)
Tue Aug 20 06:08:20 CST 2024
make: Entering directory '/usr/src/linux-headers-6.6.44-production+truenas'
make[2]: *** [scripts/Makefile.build:243: /var/lib/dkms/dxgkrnl/149cbd13f/build/dxgmodule.o] Killed
make[2]: *** Waiting for unfinished jobs....
make[2]: *** [scripts/Makefile.build:243: /var/lib/dkms/dxgkrnl/149cbd13f/build/hmgr.o] Killed
make[1]: *** [/usr/src/linux-headers-6.6.44-production+truenas/Makefile:1924: /var/lib/dkms/dxgkrnl/149cbd13f/build] Error 2
make: *** [Makefile:237: __sub-make] Error 2
make: Leaving directory '/usr/src/linux-headers-6.6.44-production+truenas
```

I'm interested in this problem. It also occurred while cross-compiling via my Debian systems.

I found that the TrueNAS 6.6 Linux kernel has an old ``eventfd_signal`` function even though it is the latest version.

Perhaps the [0002-Fix-eventfd_signal](https://github.com/staralt/dxgkrnl-dkms/blob/main/linux-msft-wsl-6.6.y/0002-Fix-eventfd_signal.patch) patch is causing this problem. Excluding this patch fixed the issue.

install.sh:
```bash
        "linux-msft-wsl-6.6.y")
            PATCHES="linux-msft-wsl-5.15.y/0001-Add-a-gpu-pv-support.patch”;
#                    linux-msft-wsl-6.6.y/0002-Fix-eventfd_signal.patch"; <———— comment out here

            for PATCH in $PATCHES; do
```

Did this fix the issue on your system?

<hr>

make.log:
```
DKMS make.log for dxgkrnl-149cbd13f for kernel 6.6.32-production+truenas (x86_64) 
Tue Aug 20 14:36:28 UTC 2024 
make: Entering directory '/usr/src/linux-headers-6.6.32-production+truenas' 
  CC [M]  /var/lib/dkms/dxgkrnl/149cbd13f/build/dxgmodule.o 
  CC [M]  /var/lib/dkms/dxgkrnl/149cbd13f/build/hmgr.o 
  CC [M]  /var/lib/dkms/dxgkrnl/149cbd13f/build/misc.o 
  CC [M]  /var/lib/dkms/dxgkrnl/149cbd13f/build/dxgadapter.o 
  CC [M]  /var/lib/dkms/dxgkrnl/149cbd13f/build/ioctl.o 
  CC [M]  /var/lib/dkms/dxgkrnl/149cbd13f/build/dxgvmbus.o 
/var/lib/dkms/dxgkrnl/149cbd13f/build/dxgmodule.c: In function 'signal_host_cpu_event': 
/var/lib/dkms/dxgkrnl/149cbd13f/build/dxgmodule.c:195:17: error: too few arguments to function 'eventfd_signal' 
  195 |                 eventfd_signal(event->cpu_event); 
      |                 ^~~~~~~~~~~~~~ 
In file included from /var/lib/dkms/dxgkrnl/149cbd13f/build/dxgmodule.c:15: 
./include/linux/eventfd.h:38:7: note: declared here 
   38 | __u64 eventfd_signal(struct eventfd_ctx *ctx, __u64 n); 
      |       ^~~~~~~~~~~~~~ 
make[2]: *** [scripts/Makefile.build:243: /var/lib/dkms/dxgkrnl/149cbd13f/build/dxgmodule.o] Error 1 
make[2]: *** Waiting for unfinished jobs.... 
make[1]: *** [/usr/src/linux-headers-6.6.32-production+truenas/Makefile:1916: /var/lib/dkms/dxgkrnl/149cbd13f/build] Error 2 
make: *** [Makefile:237: __sub-make] Error 2 
make: Leaving directory '/usr/src/linux-headers-6.6.32-production+truenas'
```

I commented it out, but have the same error.
```
DKMS make.log for dxgkrnl-149cbd13f for kernel 6.6.44-production+truenas (x86_64)
Wed Aug 21 18:45:40 CST 2024
make: Entering directory '/usr/src/linux-headers-6.6.44-production+truenas'
make[2]: *** [scripts/Makefile.build:243: /var/lib/dkms/dxgkrnl/149cbd13f/build/dxgmodule.o] Killed
make[2]: *** Waiting for unfinished jobs....
make[2]: *** [scripts/Makefile.build:243: /var/lib/dkms/dxgkrnl/149cbd13f/build/hmgr.o] Killed
make[1]: *** [/usr/src/linux-headers-6.6.44-production+truenas/Makefile:1924: /var/lib/dkms/dxgkrnl/149cbd13f/build] Error 2
make: *** [Makefile:237: __sub-make] Error 2
make: Leaving directory '/usr/src/linux-headers-6.6.44-production+truenas'
```

> I commented it out, but have the same error.
> 
> ```
> DKMS make.log for dxgkrnl-149cbd13f for kernel 6.6.44-production+truenas (x86_64)
> Wed Aug 21 18:45:40 CST 2024
> make: Entering directory '/usr/src/linux-headers-6.6.44-production+truenas'
> make[2]: *** [scripts/Makefile.build:243: /var/lib/dkms/dxgkrnl/149cbd13f/build/dxgmodule.o] Killed
> make[2]: *** Waiting for unfinished jobs....
> make[2]: *** [scripts/Makefile.build:243: /var/lib/dkms/dxgkrnl/149cbd13f/build/hmgr.o] Killed
> make[1]: *** [/usr/src/linux-headers-6.6.44-production+truenas/Makefile:1924: /var/lib/dkms/dxgkrnl/149cbd13f/build] Error 2
> make: *** [Makefile:237: __sub-make] Error 2
> make: Leaving directory '/usr/src/linux-headers-6.6.44-production+truenas'
> ```

Uh oh, I keep looking into this issue. 

Is the ``libelf-dev`` package  installed on your system?

> > 我将其注释掉，但仍然出现同样的错误。
> > ```
> > DKMS make.log for dxgkrnl-149cbd13f for kernel 6.6.44-production+truenas (x86_64)
> > Wed Aug 21 18:45:40 CST 2024
> > make: Entering directory '/usr/src/linux-headers-6.6.44-production+truenas'
> > make[2]: *** [scripts/Makefile.build:243: /var/lib/dkms/dxgkrnl/149cbd13f/build/dxgmodule.o] Killed
> > make[2]: *** Waiting for unfinished jobs....
> > make[2]: *** [scripts/Makefile.build:243: /var/lib/dkms/dxgkrnl/149cbd13f/build/hmgr.o] Killed
> > make[1]: *** [/usr/src/linux-headers-6.6.44-production+truenas/Makefile:1924: /var/lib/dkms/dxgkrnl/149cbd13f/build] Error 2
> > make: *** [Makefile:237: __sub-make] Error 2
> > make: Leaving directory '/usr/src/linux-headers-6.6.44-production+truenas'
> > ```
> 
> 哦哦，我一直在研究这个问题。
> 
> 该软件包是否`libelf-dev`安装在您的系统上？

Not installed, but after installation, the same error message still appears

--- a/README.md
+++ b/README.md
@@ -6,18 +6,12 @@ This project makes it simple to use GPU-P on Linux with the latest kernel VM. 
 
 ### Available on
 
-- Ubuntu Server 24.04 LTS (Tensorflow is working)
-   - 6.8.0-1007-azure
-   - 6.8.0-31-generic
+- Ubuntu | Ubuntu Server
+   - 24.04 LTS @ 6.8.0
+   - 22.04 LTS @ 5.15.0 ~ 6.5.0
 
-- Ubuntu 22.04 LTS (just installed, not tested)
-   - 6.5.0-18-generic
-   - 6.2.0-39-generic
-   - 6.1.0-1036-oem
-   - 6.0.0-1016-oem
-   - 5.19.0-50-generic
-   - 5.17.0-1035-oem
-   - 5.15.0-94-generic
+-  TrueNAS SCALE
+   - 23.10.2 @ 6.6.44-production+truenas
 
 ### Requirements
 

--- a/install.sh
+++ b/install.sh
@@ -28,7 +28,7 @@ install_dependencies() {
 
 update_git() {
     SYSTEM_KERNEL_VERSION="`echo ${TARGET_KERNEL_VERSION} | grep -Po ^[0-9]+\.[0-9]+`"
-    if [ "${SYSTEM_KERNEL_VERSION:0:1}" -ge "6" ] & [ "${SYSTEM_KERNEL_VERSION:2}" -ge "6" ]; then
+    if [ "${SYSTEM_KERNEL_VERSION:0:1}" -ge "6" ] && [ "${SYSTEM_KERNEL_VERSION:2}" -ge "6" ]; then
         TARGET_BRANCH="linux-msft-wsl-6.6.y";
     else
         TARGET_BRANCH="linux-msft-wsl-5.15.y";
@@ -169,7 +169,7 @@ clean() {
         exit 0
     elif [ "$1" == "all" ]; then
         TARGETS=`dkms status dxgkrnl | grep -E "dxgkrnl/[a-z0-9]+" -o | awk '!a[$0]++'`
-        if [ -z $TARGETS ]; then
+        if [ -z "$TARGETS" ]; then
             echo "Ignored. There is no modules to clean."
             exit 0
         fi

--- a/install.sh
+++ b/install.sh
@@ -8,13 +8,13 @@ install_dependencies() {
     if [[ "$LINUX_DISTRO" == *"debian"* ]]; then
         apt update;
         apt install -y git dkms;
-        if [ ! -e "/lib/modules/`uname -r`" ]; then
-            apt install -y linux-headers-`uname -r`;
+        if [ ! -e "/lib/modules/${TARGET_KERNEL_VERSION}" ]; then
+            apt install -y linux-headers-${TARGET_KERNEL_VERSION};
         fi
     elif [[ "$LINUX_DISTRO" == *"fedora"* ]]; then
         yum -y install git dkms;
-        if [ ! -e "/lib/modules/`uname -r`"]; then
-            yum -y install linux-headers-`uname -r`;
+        if [ ! -e "/lib/modules/${TARGET_KERNEL_VERSION}"]; then
+            yum -y install linux-headers-${TARGET_KERNEL_VERSION};
         fi
     else
         >&2 echo "Fatal: The system distro is unsupported";
@@ -27,7 +27,7 @@ install_dependencies() {
 }
 
 update_git() {
-    SYSTEM_KERNEL_VERSION="`uname -r | grep -Po ^[0-9]+\.[0-9]+`"
+    SYSTEM_KERNEL_VERSION="`echo ${TARGET_KERNEL_VERSION} | grep -Po ^[0-9]+\.[0-9]+`"
     if [ "${SYSTEM_KERNEL_VERSION:0:1}" -ge "6" ] & [ "${SYSTEM_KERNEL_VERSION:2}" -ge "6" ]; then
         TARGET_BRANCH="linux-msft-wsl-6.6.y";
     else
@@ -117,13 +117,20 @@ EOF
 }
 
 install_dkms() {
-    dkms add dxgkrnl/$VERSION
-    dkms build dxgkrnl/$VERSION
-    dkms install dxgkrnl/$VERSION
+    dkms -k ${TARGET_KERNEL_VERSION} add dxgkrnl/$VERSION
+    dkms -k ${TARGET_KERNEL_VERSION} build dxgkrnl/$VERSION
+    dkms -k ${TARGET_KERNEL_VERSION} install dxgkrnl/$VERSION
 }
 
 all() {
-    echo -e "\nInstalling dependencies...\n"
+    TARGET_KERNEL_VERSION="$1";
+    if [ "$TARGET_KERNEL_VERSION" == "" ]; then
+        TARGET_KERNEL_VERSION=`uname -r`
+    fi
+
+    echo -e "\nTarget Kernel Version: ${TARGET_KERNEL_VERSION}\n"
+
+    echo -e "Installing dependencies...\n"
     install_dependencies
 
     echo
@@ -139,7 +146,7 @@ all() {
 help() {
     echo
     echo "Usage:"
-    echo "  $0 - Install a latest module."
+    echo "  $0 (target kernel version) - Install a latest module."
     echo
     echo "  $0 clean all - Remove all modules."
     echo "  $0 clean [version] - Remove a specific version module."
@@ -168,23 +175,26 @@ clean() {
         fi
 
         for TARGET in $TARGETS; do
-            dkms remove "$TARGET"
+            dkms --all remove "$TARGET"
             rm -r "/usr/src/dxgkrnl-${TARGET:8}"
             echo
         done
     else
-        dkms remove "dxgkrnl/$1"
+        dkms --all remove "dxgkrnl/$1"
         rm -r "/usr/src/dxgkrnl-$1"
         echo
     fi
 }
 
 if [ -z $1 ]; then
-    all
+    all `uname -r`
 elif [ "$1" = "clean" ]; then
     shift
     clean "$@"
+elif [[ "$1" =~ ^[0-9]+\.[0-9]+\.[0-9]+.+$ ]]; then
+    all $1
 else
+    echo "Incorrect kernel version '$1' (excpeted format is '`ls /lib/modules | head -n1`')";
     help
 fi
 

--- a/install.sh
+++ b/install.sh
@@ -75,8 +75,10 @@ install() {
             done
             ;;
         "linux-msft-wsl-6.6.y")
-            PATCHES="linux-msft-wsl-5.15.y/0001-Add-a-gpu-pv-support.patch \
-                    linux-msft-wsl-6.6.y/0002-Fix-eventfd_signal.patch";
+            PATCHES="linux-msft-wsl-5.15.y/0001-Add-a-gpu-pv-support.patch";
+            if [[ "$TARGET_KERNEL_VERSION" != *"truenas"* ]]; then
+                PATCHES="$PATCHES linux-msft-wsl-6.6.y/0002-Fix-eventfd_signal.patch";
+            fi
 
             for PATCH in $PATCHES; do
                 # Patch source files

--- a/install.sh
+++ b/install.sh
@@ -1,18 +1,29 @@
 #!/bin/bash -e
 
 WORKDIR="$(dirname $(realpath $0))"
-LINUX_DISTRO="$(cat /etc/*-release | grep -Po "(?<=^ID_LIKE=).+$")"
+LINUX_DISTRO="$(cat /etc/*-release)"
+LINUX_DISTRO=${LINUX_DISTRO,,}
 
 install_dependencies() {
-    case $LINUX_DISTRO in
-        "debian"|"Debian")
-            apt install -y linux-headers-`uname -r` git dkms;;
-        "fedora"|"Fedora")
-            yum -y install linux-headers-`uname -r` git dkms;;
-        *)
-            >&2 echo "Fatal: The system distro '$LINUX_DISTRO' is unsupported"
-            exit 1;;
-    esac
+    if [[ "$LINUX_DISTRO" == *"debian"* ]]; then
+        apt update;
+        apt install -y git dkms;
+        if [ ! -e "/lib/modules/`uname -r`" ]; then
+            apt install -y linux-headers-`uname -r`;
+        fi
+    elif [[ "$LINUX_DISTRO" == *"fedora"* ]]; then
+        yum -y install git dkms;
+        if [ ! -e "/lib/modules/`uname -r`"]; then
+            yum -y install linux-headers-`uname -r`;
+        fi
+    else
+        >&2 echo "Fatal: The system distro is unsupported";
+        >&2 echo "If your system is based on 'Debian' or 'Fedora', please report this issue with the following information.";
+        >&2 echo "https://git.staralt.dev/dxgkrnl-dkms/issues";
+        >&2 echo;
+        >&2 cat /etc/*-release;
+        exit 1;
+    fi
 }
 
 update_git() {

